From ab95ff1c01df7414c4e7f6e95c1a1d9c0a584a81 Mon Sep 17 00:00:00 2001
From: Stuart MENEFY <stuart.menefy@st.com>
Date: Wed, 13 May 2009 12:37:51 +0100
Subject: [PATCH] sh_stm: Add support for ST40-300 as an explicit sub-architecture

Now that gcc supports ST40-300 as an explicit architecture variant,
add support into the kernel build. Currently this simply selects the
compilation option when building, which may allow for slightly more
optimal code generation. However it also makes CONFIG_CPU_ST40_300
available as an config option which can be used in the future.

Signed-off-by: Stuart Menefy <stuart.menefy@st.com>
---
 arch/sh/Kconfig                |    2 ++
 arch/sh/Makefile               |    3 +++
 arch/sh/kernel/cpu/sh4/probe.c |    1 +
 arch/sh/kernel/setup.c         |    2 +-
 arch/sh/mm/Kconfig             |   14 ++++++++++----
 include/asm-sh/cpu-features.h  |    3 +++
 6 files changed, 20 insertions(+), 5 deletions(-)

diff --git a/arch/sh/Kconfig b/arch/sh/Kconfig
index 7024ab0..43198a9 100644
--- a/arch/sh/Kconfig
+++ b/arch/sh/Kconfig
@@ -462,6 +462,7 @@ config SH_ST_CB101
 config SH_ST_CB102
 	bool "cb102: ST Customer board 102"
 	depends on CPU_SUBTYPE_STX7200
+	select CPU_ST40_300
 	select SND_STM_CONV_DUMMY if SND_STM
 	help
 	  Select SH_ST_CB102 if configuring for this board.
@@ -602,6 +603,7 @@ config SH_ST_MB628
 config SH_ST_MB671
 	bool "mb671: STx7200 cut 2.0 Mboard"
 	depends on CPU_SUBTYPE_STX7200
+	select CPU_ST40_300
 	select SH_ST_EPLD
 	select SH_ST_STEM
 	select SH_ST_STPM_HD_V1
diff --git a/arch/sh/Makefile b/arch/sh/Makefile
index 5fef34c..41aa2cc 100644
--- a/arch/sh/Makefile
+++ b/arch/sh/Makefile
@@ -21,6 +21,7 @@ isa-$(CONFIG_CPU_SH3)			:= sh3
 isa-$(CONFIG_CPU_SH4)			:= sh4
 isa-$(CONFIG_CPU_SH4A)			:= sh4a
 isa-$(CONFIG_CPU_SH4AL_DSP)		:= sh4al
+isa-$(CONFIG_CPU_ST40_300)		:= st40-300
 
 isa-$(CONFIG_SH_DSP)			:= $(isa-y)-dsp
 
@@ -44,6 +45,8 @@ cflags-$(CONFIG_CPU_SH4)		:= $(call cc-option,-m4,) \
 	$(call cc-option,-mno-implicit-fp,-m4-nofpu)
 cflags-$(CONFIG_CPU_SH4A)		+= $(call cc-option,-m4a,) \
 					   $(call cc-option,-m4a-nofpu,)
+cflags-$(CONFIG_CPU_ST40_300)		+= $(call cc-option,-m4-300,) \
+					   $(call cc-option,-m4-300-nofpu,)
 
 cflags-$(CONFIG_CPU_BIG_ENDIAN)		+= -mb
 cflags-$(CONFIG_CPU_LITTLE_ENDIAN)	+= -ml
diff --git a/arch/sh/kernel/cpu/sh4/probe.c b/arch/sh/kernel/cpu/sh4/probe.c
index 37e562b..6788980 100644
--- a/arch/sh/kernel/cpu/sh4/probe.c
+++ b/arch/sh/kernel/cpu/sh4/probe.c
@@ -193,6 +193,7 @@ int __init detect_cpu_and_cache_system(void)
 			break;
 		}
 		cpu_data->flags |= CPU_HAS_FPU;
+		cpu_data->flags |= CPU_HAS_ICBI | CPU_HAS_SYNCO | CPU_HAS_FPCHG;
 		ramcr = ctrl_inl(CCN_RAMCR);
 		boot_cpu_data.icache.ways = (ramcr & (1<<7)) ? 2 : 4;
 		boot_cpu_data.dcache.ways = (ramcr & (1<<6)) ? 2 : 4;
diff --git a/arch/sh/kernel/setup.c b/arch/sh/kernel/setup.c
index 891182c..ca8dad0 100644
--- a/arch/sh/kernel/setup.c
+++ b/arch/sh/kernel/setup.c
@@ -351,7 +351,7 @@ const char *get_cpu_subtype(struct sh_cpuinfo *c)
 /* Symbolic CPU flags, keep in sync with asm/cpu-features.h */
 static const char *cpu_flags[] = {
 	"none", "fpu", "p2flush", "mmuassoc", "dsp", "perfctr",
-	"ptea", "llsc", "l2", "op32", NULL
+	"ptea", "llsc", "l2", "op32", "icbi", "synco", "fpchg", NULL
 };
 
 static void show_cpuflags(struct seq_file *m, struct sh_cpuinfo *c)
diff --git a/arch/sh/mm/Kconfig b/arch/sh/mm/Kconfig
index b1c132f..6585979 100644
--- a/arch/sh/mm/Kconfig
+++ b/arch/sh/mm/Kconfig
@@ -35,6 +35,10 @@ config CPU_SUBTYPE_ST40
 	select GENERIC_GPIO
 	select HAVE_GPIO_LIB
 
+config CPU_ST40_300
+       bool
+       select CPU_SUBTYPE_ST40
+
 config CPU_SHX2
 	bool
 
@@ -178,28 +182,28 @@ config CPU_SUBTYPE_STB7100
 
 config CPU_SUBTYPE_STX5197
 	bool "Support STx5197 processors"
-	select CPU_SUBTYPE_ST40
+	select CPU_ST40_300
 	select SUPPORTS_32BIT
 	help
 	  Select STX5197 if you have an STx5197.
 
 config CPU_SUBTYPE_STX7105
 	bool "Support STx7105 processors"
-	select CPU_SUBTYPE_ST40
+	select CPU_ST40_300
 	select SUPPORTS_32BIT
 	help
 	  Select STX7105 if you have an STx7105 CPU.
 
 config CPU_SUBTYPE_STX7111
 	bool "Support STx7111 processors"
-	select CPU_SUBTYPE_ST40
+	select CPU_ST40_300
 	select SUPPORTS_32BIT
 	help
 	  Select STX7111 if you have an STx7111 CPU.
 
 config CPU_SUBTYPE_STX7141
 	bool "Support STx7141 processors"
-	select CPU_SUBTYPE_ST40
+	select CPU_ST40_300
 	select SUPPORTS_32BIT
 	help
 	  Select STX7141 if you have an STx7141 CPU.
@@ -207,6 +211,8 @@ config CPU_SUBTYPE_STX7141
 config CPU_SUBTYPE_STX7200
 	bool "Support STx7200 processor"
 	select CPU_SUBTYPE_ST40
+	# Note this doesn't select CPU_ST40_300 because cut 1 was an ST40-200.
+	# Boards designed to take cut 2 or later must set CPU_ST40_300.
 	select SUPPORTS_32BIT
 	help
 	  Select STx7200 if you have an STx7200 CPU.
diff --git a/include/asm-sh/cpu-features.h b/include/asm-sh/cpu-features.h
index 86308aa..efcd516 100644
--- a/include/asm-sh/cpu-features.h
+++ b/include/asm-sh/cpu-features.h
@@ -21,5 +21,8 @@
 #define CPU_HAS_LLSC		0x0040	/* movli.l/movco.l */
 #define CPU_HAS_L2_CACHE	0x0080	/* Secondary cache / URAM */
 #define CPU_HAS_OP32		0x0100	/* 32-bit instruction support */
+#define CPU_HAS_ICBI		0x0200	/* icbi instruction */
+#define CPU_HAS_SYNCO		0x0400	/* synco instruction */
+#define CPU_HAS_FPCHG		0x0800	/* fpchg instruction */
 
 #endif /* __ASM_SH_CPU_FEATURES_H */
-- 
1.6.0.6

